### DISCUSSION TITLE
How to bulk save with async API

### DISCUSSION THREAD

**Discussion by kigawas** [sqlalchemy/sqlalchemy#6760] [Usage Questions] [answered]
Since `session.add_all(items)` will generate multiple insert SQLs, I'd like to use bulk save in the `AsyncSession`

**Comment by zzzeek** [comment, marked as answer]
session.add_all() vs. the "bulk" methods have no difference in how INSERT is emitted to the database.

there are three levels of "INSERT".   one is that SQLAlchemy emits each row in a separate cursor.execute() call.  The next, that it emits the rows at once in a cursor.executemany() call.  The third, that it overcomes limitations in PostgreSQL drivers of "executemany()" by using the [fast execution helpers](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#orm-batch-inserts-with-psycopg2-now-batch-statements-with-returning-in-most-cases).

To use executemany(), just make sure that your objects have their primary key set up before the session flushes them; executemany() will be used.   To use the fast execution helpers, these are only implemented for psycopg2 right now so are not available within asyncio yet, there's an issue to implement this for asyncpg.

to have complete control over how to insert, use [Core](https://docs.sqlalchemy.org/en/14/tutorial/data_insert.html).  to insert the row as one big VALUES statement, which is vastly more performant on PostgreSQL, use [insert with multiple values](https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Insert.values.params.*args).

Finally, you can of course use bulk save in asyncio by using [run_sync](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#running-synchronous-methods-and-functions-under-asyncio), like `await async_session.run_sync(lambda session: session.bulk_save(...))`.
